### DISCUSSION TITLE
What's the deal with tab sizes on GitHub Gists?

### DISCUSSION THREAD

**Discussion by s-h-a-d-o-w** [community/community#23860] [New to GitHub]
<p>I just created my first gist and set the Tab size to 4 but once I saved it, it got rendered with 8 spaces for some reason.</p>
<p>Horrified about the look, I hit “edit”, saw that it was set to 8, set it back to 4, saved it and again - 8…</p>
<p>If GitHub simply doesn’t support different Tab indentations (which I realize has been requested for years: <a href="https://github.com/isaacs/github/issues/170" rel="nofollow noopener">https://github.com/isaacs/github/issues/170</a> ), then at least that drop down box for size should be greyed out when “Tabs” is selected…</p>
<p>But of course I’m hoping that this is all just some… misunderstanding and that there is in fact a way to make proper tab rendering work because… it’s kind of a trivial thing.</p>

**Comment by ThirdEyeRose** [comment]
<p>Hi <a class="mention" href="/s-h-a-d-o-w">@s-h-a-d-o-w</a>,</p>
<p>The Gist tab size settings only applies to the editor and not to the rendered file. The default tab size is 8. There is no way to permanently change that, however you can modify it for a specific page load by using the <code>ts=&lt;size&gt;</code> URL parameter.</p>
<p>Thanks!</p>

**Comment by s-h-a-d-o-w** [comment]
<p>What a shame (and strange that that URL parameter exists but no UI display setting that does that) but thanks for the confirmation!</p>

**Comment by s-h-a-d-o-w** [comment]
<p>Actually, in the case someone else stumbles across this - I got aware that there are workarounds in the form of extensions for Chrome available (e.g. GitHub Custom Tab Size) and this inspired me to develop my own one that integrates directly into GitHub:</p>
<p><a href="https://github.com/s-h-a-d-o-w/github-userscripts" rel="nofollow noopener">https://github.com/s-h-a-d-o-w/github-userscripts</a></p>

**Comment by esr360** [comment]
<p>“There is no way to permanently change that”</p>
<p>Wrong - there is no way for us users to permanently change it, but Github obviously has the ability to change it from their end, which is something that shouldn’t be dismissed. 8 is far too big and ugly for the default size.</p>

**Comment by Skaruts** [comment]
<p>Why on earth would a code repository default to 8 spaces to begin with, in a world where no one uses 8 spaces, is beyond me. Not blaming you or anyone in particular. I’m just saying. It just makes no sense.</p>
<p>It’s a matter of preference, and I’m fine with anyone prefering 8 spaces, but then it also makes no sense that such a basic option to suit the user’s preference isn’t available to the user. Especially after all these years.</p>
<p>Currently the only workaround is to go through the trouble of switching indentation to spaces in your editor/IDE of choice before comitting stuff to Github… </p>

**Comment by matthewjumpsoffbuildings** [comment]
<p>Obviously someone high up at github really really hates tabs, and is trying to make it as ugly and annoying as possible for those who prefer them.</p>
<p>Seriously that is the only possible reason I can come up with as to why this absolutely insane tab size is not only offered, but default, and impossible to permanently override from the UI.</p>
<p>What the hell github?</p>

**Reply by lveillard** [reply]
This is the only explanation that makes sense

**Comment by tychop** [comment]
<p>True. It must be a tab hater. They purposly set the tab width to 8 just to annoy us tabbies. Do they really not see that you can setup tabs to use as much space as 4 spaces?</p>

**Comment by eggbean** [comment]
<p>I found a workaround/solution:</p>
<p><a href="https://github.com/isaacs/github/issues/170#issuecomment-150489692" rel="nofollow noopener">https://github.com/isaacs/github/issues/170#issuecomment-150489692</a></p>
<p>Cheers.</p>

**Comment by phpbbireland** [comment]
<p>Been waiting for this to be fixed since Adam was a boy and still not fixed…</p>
<p>You can change to 4 in setting but it defaults back to 8…<br>
Code looks terrible with all those spaces, who in the world uses 8 spaces in tab?</p>

**Comment by bradkrane** [comment]
<p>You should be able to specify the rendered tabs and spaces. An intern should be able to code this, maybe that’s a stretch, but why is it impossible to set the tab spacing at a reasonable, or any specified length for rendering after the fact? Auto converting to tab =&gt; 8 fails at WUSIWUG. I cant paste code with space: 2 for tabs and it screw up the rendering with the converted to tab: 8, what, really?</p>
<p>Come on MS. Or does gisthub have a github for a pull request for this what should already be baked in feature request?</p>

**Comment by bradkrane** [comment]
<p>Looking at the case more closely we need to open a new issue this one is dead.</p>
<p>It is not-so clearly, but suspected this issue is ‘closed’ and future updates &gt; /dev/null as this has been ‘SOLVED BY THAT-PAT IN POST <span class="hashtag">#2</span>’</p>
<p>Hence no new attention despite the additional responses, at least I hope this to be true…</p>
<p>Re-escalate the issue <a href="https://github.community/t/why-can-tab-spacing-not-be-set-in-gists/175511" class="inline-onebox">Why can tab spacing not be set in gists?</a></p>
<p>And to be sure this is not a solution: <a href="https://github.com/isaacs/github/issues/170#issuecomment-150489692" class="inline-onebox" rel="noopener nofollow ugc">pre { tab-size: 4 } · Issue #170 · isaacs/github · GitHub</a></p>
<p>There should be a UI option and per-file not what, globally?</p>

**Comment by rockwotj** [comment, marked as answer]
<p>This seems to be a user setting now: <a href="https://github.blog/changelog/2021-09-21-tab-size-rendering-preference/" class="inline-onebox" rel="noopener nofollow ugc">Tab size rendering preference | GitHub Changelog</a></p>

**Reply by NatoBoram** [reply]
It doesn't work in pull requests

**Comment by crozone** [comment]
Not sure if this is the right issue, but GitHub's current solution for changing the tab width on code only works for the values 1 to 12 anyway.

If you want something wider, like 16 or 24 (which is useful for column based files like assembly listings, tab separated CSV, etc) it simply won't work.

I've traced the issue to the `tab-size` class here: (I'm not sure where the actual source code is for this generated asset)

https://github.githubassets.com/assets/github-762fdeac7c4d.css

The relevant CSS is:

```
.tab-size[data-tab-size="1"]{tab-size:1}.tab-size[data-tab-size="2"]{tab-size:2}.tab-size[data-tab-size="3"]{tab-size:3}.tab-size[data-tab-size="4"]{tab-size:4}.tab-size[data-tab-size="5"]{tab-size:5}.tab-size[data-tab-size="6"]{tab-size:6}.tab-size[data-tab-size="7"]{tab-size:7}.tab-size[data-tab-size="8"]{tab-size:8}.tab-size[data-tab-size="9"]{tab-size:9}.tab-size[data-tab-size="10"]{tab-size:10}.tab-size[data-tab-size="11"]{tab-size:11}.tab-size[data-tab-size="12"]{tab-size:12}
```

As you can see they only define a mapping from the `data-tab-size` attribute to the css `tab-size` for the values 1..12. Github refuses to output a `data-tab-size` value outside this range, eg if you set the `?ts=` query parameter > 12, so it's clearly intentional.

**Comment by katiehuangx** [comment]
I came here because I'm still having this issue after setting my tab size preference to 2. The tab size on the editor overwrote my settings and switched back to tab size 8. 

This is driving me crazy.

**Reply by eggbean** [reply]
It's possible that you have an `.editorconfig` file in your repo that is overriding the setting?

**Reply by katiehuangx** [reply]
Nope, I don't have any other files apart from `.md`s in my repo. I found a temporary solution by writing the markdown with code in VS Code and then copying and pasting it back into GitHub.

**Comment by Foxfire1st** [comment]
that in 2025 this somehow is such a big issue... just tell people to use anything else and move on. Waste of time.
